File: src/vost/cli/_mirror.py
"""Backup and restore commands."""

from __future__ import annotations

import click

from ._helpers import (
    main,
    _repo_option,
    _dry_run_option,
    _no_create_option,
    _require_repo,
    _status,
    _open_store,
    _open_or_create_bare,
)


# ---------------------------------------------------------------------------
# Helpers
# ---------------------------------------------------------------------------

def _print_diff(diff, direction: str) -> None:
    """Pretty-print a MirrorDiff to stdout."""
    verb = "push" if direction == "push" else "pull"
    if diff.in_sync:
        click.echo(f"Nothing to {verb} — already in sync.")
        return
    for c in sorted(diff.add, key=lambda c: c.ref):
        click.echo(f"  create  {c.ref}  {c.new_target[:7]}")
    for c in sorted(diff.update, key=lambda c: c.ref):
        click.echo(f"  update  {c.ref}  {c.old_target[:7]} -> {c.new_target[:7]}")
    for c in sorted(diff.delete, key=lambda c: c.ref):
        click.echo(f"  delete  {c.ref}  {c.old_target[:7]}")
    click.echo(f"{diff.total} ref(s) would be changed.")


def _parse_refs(ref_tuple):
    """Parse --ref values into a list or dict.

    If any value contains ``:``, all values are parsed as ``src:dst``
    and a dict is returned.  Otherwise a plain list is returned.
    ``None`` is returned when the tuple is empty.
    """
    if not ref_tuple:
        return None
    has_rename = any(":" in r for r in ref_tuple)
    if not has_rename:
        return list(ref_tuple)
    result = {}
    for r in ref_tuple:
        if ":" in r:
            src, _, dst = r.partition(":")
            if not src or not dst:
                raise click.BadParameter(
                    f"Invalid ref rename {r!r} — expected src:dst",
                    param_hint="'--ref'",
                )
            result[src] = dst
        else:
            result[r] = r
    return result


def _progress_cb(ctx):
    """Return a progress callback if verbose mode is on, else None."""
    if not ctx.obj.get("verbose"):
        return None
    def _on_progress(msg):
        text = msg.decode()
        text = text.replace("\r", "\r\033[K")
        click.echo(text, nl=False, err=True)
    return _on_progress


# ---------------------------------------------------------------------------
# backup
# ---------------------------------------------------------------------------

@main.command("backup")
@_repo_option
@click.argument("url")
@_dry_run_option
@click.option("--ref", multiple=True, help="Ref to include (repeatable). Use src:dst to rename. Omit for all refs.")
@click.option("--format", "fmt", type=click.Choice(["bundle"]), default=None,
              help="Force output format (auto-detected from .bundle extension).")
@click.option("--squash", is_flag=True, default=False,
              help="Strip history: each ref becomes a single parentless commit (bundle only).")
@click.pass_context
def backup_cmd(ctx, url, dry_run, ref, fmt, squash):
    """Push refs to a remote URL or write a bundle file.

    Without --ref this is a full mirror: remote-only refs are deleted.
    With --ref only the specified refs are pushed (no deletes).
    Use --ref src:dst to rename refs on the remote side.

    If URL ends with .bundle, a portable bundle file is written.
    Use --squash to strip history from bundle output.
    """
    from ..mirror import _is_bundle_path, resolve_credentials

    store = _open_store(_require_repo(ctx))
    refs = _parse_refs(ref)
    use_bundle = (fmt == "bundle") or _is_bundle_path(url)
    auth_url = url if use_bundle else resolve_credentials(url)
    diff = store.backup(auth_url, dry_run=dry_run, progress=_progress_cb(ctx),
                        refs=refs, format=fmt, squash=squash)
    if dry_run:
        _print_diff(diff, "push")
    else:
        _status(ctx, f"Backed up to {url}")


# ---------------------------------------------------------------------------
# restore
# ---------------------------------------------------------------------------

@main.command("restore")
@_repo_option
@click.argument("url")
@_dry_run_option
@_no_create_option
@click.option("--ref", multiple=True, help="Ref to include (repeatable). Use src:dst to rename. Omit for all refs.")
@click.option("--format", "fmt", type=click.Choice(["bundle"]), default=None,
              help="Force input format (auto-detected from .bundle extension).")
@click.pass_context
def restore_cmd(ctx, url, dry_run, no_create, ref, fmt):
    """Fetch refs from a remote URL or import a bundle file.

    Restore is additive: refs are added and updated but local-only
    refs are never deleted. Use --ref src:dst to rename refs locally.
    HEAD (the current branch) is not restored;
    use 'vost branch current -b NAME' afterwards if needed.
    """
    from ..mirror import _is_bundle_path, resolve_credentials

    repo_path = _require_repo(ctx)
    store = _open_store(repo_path) if no_create else _open_or_create_bare(repo_path)
    refs = _parse_refs(ref)
    use_bundle = (fmt == "bundle") or _is_bundle_path(url)
    auth_url = url if use_bundle else resolve_credentials(url)
    diff = store.restore(auth_url, dry_run=dry_run, progress=_progress_cb(ctx),
                         refs=refs, format=fmt)
    if dry_run:
        _print_diff(diff, "pull")
    else:
        _status(ctx, f"Restored from {url}")

File: rs/src/cli/mirror.rs
use clap::Args;

use crate::types::{BackupOptions, MirrorDiff, RestoreOptions};

use super::error::CliError;
use super::helpers::*;

#[derive(Args, Debug)]
pub struct BackupArgs {
    /// Destination URL, local path, or bundle file path.
    pub url: String,
    /// Show what would change without writing.
    #[arg(short = 'n', long)]
    pub dry_run: bool,
    /// Ref to include (repeatable). Use src:dst to rename.
    #[arg(long = "ref")]
    pub refs: Vec<String>,
    /// Force output format.
    #[arg(long, value_parser = ["bundle"])]
    pub format: Option<String>,
    /// Strip history from bundle output.
    #[arg(long)]
    pub squash: bool,
}

#[derive(Args, Debug)]
pub struct RestoreArgs {
    /// Source URL, local path, or bundle file path.
    pub url: String,
    /// Show what would change without fetching.
    #[arg(short = 'n', long)]
    pub dry_run: bool,
    /// Do not auto-create the repository.
    #[arg(long)]
    pub no_create: bool,
    /// Ref to include (repeatable). Use src:dst to rename.
    #[arg(long = "ref")]
    pub refs: Vec<String>,
    /// Force input format.
    #[arg(long, value_parser = ["bundle"])]
    pub format: Option<String>,
}

fn parse_refs(
    ref_args: &[String],
) -> Result<Option<ParsedRefs>, CliError> {
    if ref_args.is_empty() {
        return Ok(None);
    }
    let has_rename = ref_args.iter().any(|r| r.contains(':'));
    if !has_rename {
        return Ok(Some(ParsedRefs::List(
            ref_args.iter().cloned().collect(),
        )));
    }
    let mut map = std::collections::HashMap::new();
    for r in ref_args {
        if let Some((src, dst)) = r.split_once(':') {
            if src.is_empty() || dst.is_empty() {
                return Err(CliError::new(format!(
                    "Invalid ref rename '{}' — expected src:dst",
                    r
                )));
            }
            map.insert(src.to_string(), dst.to_string());
        } else {
            map.insert(r.clone(), r.clone());
        }
    }
    Ok(Some(ParsedRefs::Map(map)))
}

enum ParsedRefs {
    List(Vec<String>),
    Map(std::collections::HashMap<String, String>),
}

fn is_bundle_path(url: &str) -> bool {
    url.to_lowercase().ends_with(".bundle")
}

fn print_diff(diff: &MirrorDiff, direction: &str) {
    let verb = if direction == "push" {
        "push"
    } else {
        "pull"
    };
    if diff.in_sync() {
        println!("Nothing to {} — already in sync.", verb);
        return;
    }
    let mut changes: Vec<_> = diff.add.iter().map(|c| ("create", c)).collect();
    changes.extend(diff.update.iter().map(|c| ("update", c)));
    changes.extend(diff.delete.iter().map(|c| ("delete", c)));
    changes.sort_by(|a, b| a.1.ref_name.cmp(&b.1.ref_name));

    for (action, c) in &changes {
        match *action {
            "create" => {
                println!(
                    "  create  {}  {}",
                    c.ref_name,
                    &c.new_target.as_deref().unwrap_or("")[..7.min(c.new_target.as_deref().unwrap_or("").len())]
                );
            }
            "update" => {
                println!(
                    "  update  {}  {} -> {}",
                    c.ref_name,
                    &c.old_target.as_deref().unwrap_or("")[..7.min(c.old_target.as_deref().unwrap_or("").len())],
                    &c.new_target.as_deref().unwrap_or("")[..7.min(c.new_target.as_deref().unwrap_or("").len())]
                );
            }
            "delete" => {
                println!(
                    "  delete  {}  {}",
                    c.ref_name,
                    &c.old_target.as_deref().unwrap_or("")[..7.min(c.old_target.as_deref().unwrap_or("").len())]
                );
            }
            _ => {}
        }
    }
    println!("{} ref(s) would be changed.", diff.total());
}

pub fn cmd_backup(
    repo_path: &str,
    args: &BackupArgs,
    verbose: bool,
) -> Result<(), CliError> {
    let store = open_store(repo_path)?;
    let parsed = parse_refs(&args.refs)?;

    let use_bundle =
        args.format.as_deref() == Some("bundle") || is_bundle_path(&args.url);
    let url = if use_bundle {
        args.url.clone()
    } else {
        crate::mirror::resolve_credentials(&args.url)
    };

    let mut opts = BackupOptions {
        dry_run: args.dry_run,
        squash: args.squash,
        format: args.format.clone(),
        ..Default::default()
    };

    match parsed {
        Some(ParsedRefs::List(list)) => {
            opts.refs = Some(list);
        }
        Some(ParsedRefs::Map(map)) => {
            opts.ref_map = Some(map);
        }
        None => {}
    }

    let diff = store.backup(&url, &opts).map_err(CliError::from)?;

    if args.dry_run {
        print_diff(&diff, "push");
    } else {
        status(verbose, &format!("Backed up to {}", args.url));
    }
    Ok(())
}

pub fn cmd_restore(
    repo_opt: &Option<String>,
    args: &RestoreArgs,
    verbose: bool,
) -> Result<(), CliError> {
    let repo_path = require_repo(repo_opt)?;
    let store = if args.no_create {
        open_store(&repo_path)?
    } else {
        open_or_create_bare(&repo_path)?
    };

    let parsed = parse_refs(&args.refs)?;

    let use_bundle =
        args.format.as_deref() == Some("bundle") || is_bundle_path(&args.url);
    let url = if use_bundle {
        args.url.clone()
    } else {
        crate::mirror::resolve_credentials(&args.url)
    };

    let mut opts = RestoreOptions {
        dry_run: args.dry_run,
        format: args.format.clone(),
        ..Default::default()
    };

    match parsed {
        Some(ParsedRefs::List(list)) => {
            opts.refs = Some(list);
        }
        Some(ParsedRefs::Map(map)) => {
            opts.ref_map = Some(map);
        }
        None => {}
    }

    let diff = store.restore(&url, &opts).map_err(CliError::from)?;

    if args.dry_run {
        print_diff(&diff, "pull");
    } else {
        status(verbose, &format!("Restored from {}", args.url));
    }
    Ok(())
}
